src/widgets/call-panel.jsx

```diff
@@ -0,0 +1,396 @@
+        break;
+      default:
+        console.log("Error opening your camera and/or microphone: " + e.message);
+        break;
+    }
+
+    // Make sure we shut down our end of the RTCPeerConnection so we're
+    // ready to try again.
+    this.handleCloseClick();
+  }
+
+  handleVideoOfferMsg(info) {
+    var localStream = null;
+
+    const pc = this.createPeerConnection();
+
+    var desc = new RTCSessionDescription(info.payload);
+
+    pc.setRemoteDescription(desc).then(() => {
+      console.log('!!! handleVideoOffer/setRemoteDescription -> getUserMesia');
+      return navigator.mediaDevices.getUserMedia(this.localStreamConstraints);
+    })
+    .then(stream => {
+      localStream = stream;
+      this.localRef.current.srcObject = stream;
+      this.setState({localStream: stream});
+
+      localStream.getTracks().forEach(track => {
+        console.log('local stream A adding track ', track.id, track.kind, track.readyState);
+        pc.addTrack(track, localStream);
+      });
+    })
+    .then(() => {
+      return pc.createAnswer();
+    })
+    .then(answer => {
+      return pc.setLocalDescription(answer);
+    })
+    .then(() => {
+      this.props.onSendAnswer(this.props.topic.name, this.props.seq, pc.localDescription.toJSON());
+    })
+    .catch(this.handleGetUserMediaError);
+  }
+
+  handleRemoteHangup() {
+    this.handleCloseClick();
+  }
+
+  stopTracks(el) {
+    if (el == null) { return; }
+    let stream = el.srcObject;
+    if (stream == null) { return; }
+    let tracks = stream.getTracks();
+
+    if (tracks != null) {
+      tracks.forEach((track) => {
+        track.stop();
+        track.enabled = false;
+        console.log('stopped track ', track.id, track.kind, track.readyState);
+      });
+    }
+    el.srcObject = null;
+    el.src = '';
+  }
+
+  stop() {
+    this.stopTracks(this.localRef.current);
+    this.stopTracks(this.remoteRef.current);
+    if (this.state.pc != null) {
+      this.state.pc.ontrack = null;
+      this.state.pc.onremovetrack = null;
+      this.state.pc.onremovestream = null;
+      this.state.pc.onicecandidate = null;
+      this.state.pc.oniceconnectionstatechange = null;
+      this.state.pc.onsignalingstatechange = null;
+      this.state.pc.onicegatheringstatechange = null;
+      this.state.pc.onnegotiationneeded = null;
+
+      this.state.pc.close();
+    }
+    this.setState({pc: null});
+  }
+
+  handleCloseClick() {
+    this.stop();
+    this.props.onHangup(this.props.topic.name, this.props.seq);
+  }
+
+  // Mute/unmute audio/video (specified by kind).
+  toggleMedia(kind) {
+    const stream = this.state.localStream;
+    stream.getTracks().forEach(track => {
+      if (track.kind != kind) {
+        return;
+      }
+      console.log('local stream: toggling track ', track.id, track.kind, track.readyState);
+      track.enabled = !track.enabled;
+    });
+    // Make sure we redraw the UI properly.
+    this.forceUpdate();
```

Comment:
What exactly needs to be redrawn here? If neither props nor state changes, the UI should not change.